scripts/generate_hamilton_input_qpcr_dilution.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python
+import re
+import sys
+
+from EPPs.common import GenerateHamiltonInputEPP, InvalidStepError
+
+
+class GenerateHamiltonInputQPCRDilution(GenerateHamiltonInputEPP):
+    # define the column headers that will be used in the Hamilton input file and add to the csv_array to be
+    # used to write the file
+    csv_column_headers = ['Input Plate', 'Input Well', 'Output Plate', 'Output Well', 'Sample Volume', 'Buffer Barcode',
+                          'Buffer Volume']
+    # Define the output file
+    output_file_name = 'QPCR DILUTION.csv'
+
+    # Define the number of input containers that are permitted
+    _max_nb_input_containers = 1
+
+    # Define the number of output containers that are permitted
+    _max_nb_output_containers = 1
+
+    # create the set to hold the inputs to be updated with the output from the calculation
+    artifacts_to_update = set()
+    # input udfs to be transposed to output udfs
+    udf_names = ['Adjusted Conc. (nM)', 'Ave. Conc. (nM)', 'Original Conc. (nM)', '%CV']
+
+    def calculate_rsb_volume(self, input_art):
+        """
+        The purpose of this function is to calculate the required volume of buffer to be added to a set volume of a DCT sample
+        to achieve a target concentration. The dilution will only occur if the DCT sample concentration is above a threshold
+         concentration. The script will also populate the output udfs of the step with a number input artifact udfs that
+         will get overwritten when the QPCR is re-run. This means the values are permanently stored and can be retried in
+          future.
+        """
+        output = self.process.outputs_per_input(input_art.id, ResultFile=True)[0]
+
+        for name in self.udf_names:
+            try:
+                output.udf[name] = input_art.udf[name]
+            except:
```

Comment:
We should not use except without any Exception to catch. Here you probably want to catch KeyError if a udf is missing. Also that will not check that the value in the UDFs are floats.
Something like this would be more explicit:
```python
if name in input_art.udf.get(name) is not None:
    output.udf[name] = input_art.udf[name]
else:
    raise InvalidStepError('UDF population failed due to missing input value in %s' % name)
```

Reply:
Updated